### PR TITLE
correct EBAS data reading 

### PR DESCRIPTION
**Pull request by jgriesfeller**
1st try to solve #592

### CODE REVIEW

**Comment by jgriesfeller** [comment]
I checked the test data. The sqlite DB that is part of the test data set does not contain the table needed for the changes I made. So we need to curate the EBAS test data from scratch (new DB do not fit elder file lists since the file names change over time). 
I hope to find some kind of preparation script for that (since we in principle should update the test data with every data update we get from Nilu to take into account all possible changes), but I am not that optimistic that something like that exists.

**Comment by jgriesfeller** [comment]
Found the testdata prep scrips in `~/MyPyaerocom/testdata-minimal/scripts` but in no repository so far. They are not in https://github.com/metno/pyaerocom-dev-scripts.
I think we should include them in the standard pyaerocom repository e.g. in `tests/scripts`. They should be easy to find and easy to maintain and I don't think they contain info we do not want to share.
@avaldebe What's your opnion on that?

**Comment by jgriesfeller** [comment]
EBAS CI is finally working again.
Unfortunately the needed change to make the `concco` reading work lengthens the CI run time by roughly 5 minutes due to the usage of another database table in addition.
The new EBAS data is a bit different than before as it adds many data in units of `ug m-3` in addition to e.g. `ug S m-3` or `ug C m-3`. In some cases this leads to reading errors because we do an automatic conversion from e.g. `ug S m-3` to `ug m-3` based on the code here: https://github.com/metno/pyaerocom/blob/d42d66d86d99de7c75b174f3c2614cdbc675bcb6/pyaerocom/units_helpers.py#L40-L56 

@hildefagerli and I therefore decided remove the automatic conversion and make it possible to read these variables in e.g. `ug S m-3` under different variable name. Unfortunately this is non trivial because variable name aliasing also changes the desired unit to the one of the alias variable name. So that change will in a separate PR.

Due to some errors in the current EBAS extract, this PR probaly should only be merged when we have a corrected EBAS extract.

**this PR will break EBAS reading for EBAS data revisions from before 2022!**

File: tests/scripts/create_subset_ebas.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+    if not CURRENT_OK:
+        print(f"outdated files: {files_invalid}")
+        print("---------------------WARNING------------------------------")
+        print(
+            "Some files in current EBAS subset do not exist anymore in "
+            "the most recent EBAS database, consider updating them"
+        )
+        print("---------------------WARNING------------------------------")
+        raise Exception
```
Comment (avaldebe) [inline tests/scripts/create_subset_ebas.py:182]: Is this an exception or a warning?
I think this should be

``` python
    if not CURRENT_OK:
        raise RuntimeError(
            "Some files in current EBAS subset do not exist anymore in "
            "the most recent EBAS database. "
            f"Outdated files: {files_invalid}"
        )
```

File: tests/scripts/create_subset_ebas.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+                except Exception as e:
+                    print(e)
+
+    else:
+        infofile = JSON_FILE
+        if os.path.exists(infofile):
+            with open(infofile, "r") as f:
+                current_files = simplejson.load(f)
+        else:
+            current_files = {}
```
Comment (avaldebe) [inline tests/scripts/create_subset_ebas.py:218]: `JSON_FILE` is a Path object, so this section can be written as

``` python
        if JSON_FILE.exists():
            current_files = simplejson.loads(JSON_FILE.read_text())
        else:
            current_files = {}
```

File: tests/scripts/create_subset_ebas.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                files, overlap_found = get_files_var_statlist(data, var, STATSFIX, overlap_found)
+            else:
+                files = get_files_var_statnum(data, var, VARS_STATVAR_NUM)
+            all_files[var] = files
+
+        if not UPDATE:
+            print("NOTHING WILL BE COPIED TO TEST DATA")
+        else:
+
+            src = Path(EBAS_BASE_DIR).joinpath("data")
```
Comment (avaldebe) [inline tests/scripts/create_subset_ebas.py:249]: `EBAS_BASE_DIR` was upgraded to a `Path` object on a previous suggestion, so this libe becomes

``` python
            src = EBAS_BASE_DIR / "data"
```

File: tests/scripts/create_subsets_aeronet.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        statnames = []
+
+        for name, data in loaded.items():
+
+            subset = data.apply_filters(**{attr: val})
+
+            subsets[name] = subset
+            statnames.append(subset.unique_station_names)
+
+        # same_stats = np.intersect1d(*statnames)
```
Comment (avaldebe) [inline tests/scripts/create_subsets_aeronet.py:62]: please remove dead code

File: tests/scripts/create_subsets_aeronet.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        same_stats = reduce(np.intersect1d, statnames)
```
Comment (avaldebe) [inline tests/scripts/create_subsets_aeronet.py:64]: what is this?

File: tests/scripts/create_subsets_aeronet.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                    break
+                data = stat[var]
+                numvalid = np.sum(~np.isnan(data))
+                if not numvalid > MIN_NUM_VALID:
+                    statok = False
+                    break
+            if statok:
+                stats_ok.append(statname)
+        if len(stats_ok) == 0:
+            raise Exception
```
Comment (avaldebe) [inline tests/scripts/create_subsets_aeronet.py:89]: please provide an exception message, e.g.

``` python
        if len(stats_ok) == 0:
            raise RuntimeError("no stats were computed")
```

File: tests/scripts/create_subsets_ghost.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+Created on Fri Feb 26 09:17:09 2021
+
+@author: jonasg
+"""
+import os
+
+import matplotlib.pyplot as plt
+
+plt.close("all")
```
Comment (avaldebe) [inline tests/scripts/create_subsets_ghost.py:14]: is this reevant?

**Review by avaldebe — CHANGES_REQUESTED** [review]
If I understand correctly the `tests/scripts` are the scripts needed to create and package the `testdata-minimal.tar.gz` file.

They are not tests, fixtures or test datasets, so they should not be on `tests`. I suggests to move them to `scripts/testdata-minimal`.

File: pyaerocom/access_testdata.py [hunk 8]
```diff
@@ -24,7 +24,7 @@
 class AccessTestData:
 
     #: That's were the testdata can be downloaded from
-    URL_TESTDATA = "https://pyaerocom.met.no/pyaerocom-suppl/testdata-minimal.tar.gz"
+    URL_TESTDATA = "https://pyaerocom.met.no/pyaerocom-suppl/testdata-minimal.tar.gz.ebas_202201"
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply pyaerocom/access_testdata.py:27]: the extension is not important on Unix like OSes. At some point I want to go back to the original name anyway, but for testing I needed a different name

File: pyaerocom/io/ebas_file_index.py [hunk 9]
```diff
@@ -112,7 +112,13 @@ def make_file_query_str(self, distinct=True, **kwargs):
...
             SQL file request command for current specs
         """
-        return self.make_query_str(distinct=distinct, **kwargs)
+        query = self.make_query_str(distinct=distinct, **kwargs)
+        # add an extsion to get only files that have no fraction variables in them
+        query = query.replace(
+            ";",
+            " and not exists (select * from characteristic where var_id=variable.var_id and ct_type='Fraction');",
+        )
+        return query
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply pyaerocom/io/ebas_file_index.py:121]: no as there are cases where the extension is neither needed nor will work. It's only needed if one asks for the files, but there's plenty of other queries in the code

File: pyaerocom/io/ebas_file_index.py [hunk 10]
```diff
@@ -305,10 +311,18 @@ def execute_request(self, request):
...
         if isinstance(request, EbasSQLRequest):
-            request = request.make_query_str()
+            if not file_request:
+                sql_str = request.make_query_str()
+            else:
+                sql_str = request.make_file_query_str()
+        else:
+            sql_str = request
+
+        print(sql_str)
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply pyaerocom/io/ebas_file_index.py:321]: sorry leftover from testing

File: pyaerocom/access_testdata.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply pyaerocom/access_testdata.py:27]: The moving of the scripts is a good suggestion

**Comment by jgriesfeller** [comment]
The whole EBAS reading is quite a hack.
Because Nilu just supplied an updates dataset with yet another data format change, I have to reiterate the whole stuff once again.

**Comment by jgriesfeller** [comment]
adding a comment to myself:
https://github.com/metno/pyaerocom/blob/0be255151d723da7952ca619b5b7a34369580f65/pyaerocom/units_helpers.py#L40-L59 is used for variable conversion on observations **and** modeldata. 
Due to the way the EBAS reading works (trying to convert every possible unit, but expecting just one variable in a file to match) this is not working. I either have to split these unit conversion function into one for observations and one for models or fix the variable matching. The former is easier but messier unfortunately

File: tests/io/test_ebas_file_index.py [hunk 11]
```diff
@@ -35,11 +35,15 @@ def test_EbasSQLRequest__var2sql_error():
...
             {},
-            "select distinct filename from variable join station on station.station_code=variable.station_code;",
+            # "select distinct filename from variable;",
+            "select distinct filename from variable join station on station.station_code=variable.station_code and not exists (select * from characteristic where var_id=variable.var_id and ct_type='Fraction');",
+            # "select distinct filename from variable join station on station.station_code=variable.station_code;",
         ),
         (
             {"distinct": False},
-            "select filename from variable join station on station.station_code=variable.station_code;",
+            # "select filename from variable ;",
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply tests/io/test_ebas_file_index.py:44]: done

File: tests/scripts/calc_example_coldata.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+OUTBASE = TESTDATADIR.joinpath("coldata")
+
+if not OUTBASE.exists():
+    OUTBASE.mkdir()
+
+fpath = TESTDATADIR.joinpath(CHECK_PATHS["tm5aod"])
+if not fpath.exists():
+    raise Exception("Unexpected error, please debug")
+mod = pya.GriddedData(fpath)
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply tests/scripts/calc_example_coldata.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subset_ebas.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+
+
+tda = AccessTestData()
+
+TESTDATADIR = tda.basedir
+
+OUTBASE = Path(TESTDATADIR).joinpath("testdata-minimal/obsdata/EBASMultiColumn")
+SCRIPT_BASE_DIR = Path(TESTDATADIR).joinpath("testdata-minimal/scripts")
+
+FILES_DEST = OUTBASE.joinpath("data")
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply tests/scripts/create_subset_ebas.py]: `SCRIPT_BASE_DIR` is correct, but yesy it's a bit different to the other folder structure due to the tar we get as data.
The structure will be adjusted with #604

File: tests/scripts/create_subset_ebas.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+SEARCH_PROBLEM_FILES = False
+NAME = "EBASMC"
+
+# if ebas_local is not None:
+#     FILES_SRC = ebas_local
+# else:
+EBAS_BASE_DIR = "/lustre/storeA/project/aerocom/aerocom1/AEROCOM_OBSDATA/EBASMultiColumn/data/"
+assert os.path.exists(EBAS_BASE_DIR)
+
+JSON_FILE = SCRIPT_BASE_DIR.joinpath("ebas_files.json")
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply tests/scripts/create_subset_ebas.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subset_ebas.py (hunk 1 above)
Comment (jgriesfeller) [reply tests/scripts/create_subset_ebas.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subset_ebas.py (hunk 3 above)
Comment (jgriesfeller) [reply tests/scripts/create_subset_ebas.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subsets_aeronet.py (hunk 5 above)
Comment (jgriesfeller) [reply tests/scripts/create_subsets_aeronet.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subsets_ghost.py (hunk 7 above)
Comment (jgriesfeller) [reply tests/scripts/create_subsets_ghost.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subset_ebas.py (hunk 2 above)
Comment (jgriesfeller) [reply tests/scripts/create_subset_ebas.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/scripts/create_subsets_aeronet.py (hunk 4 above)
Comment (jgriesfeller) [reply tests/scripts/create_subsets_aeronet.py]: I put that into a separate issue #604. A bit out of scope for now.

File: tests/io/test_ebas_file_index.py [hunk 15]
```diff
@@ -35,11 +35,15 @@ def test_EbasSQLRequest__var2sql_error():
     [
         (
             {},
-            "select distinct filename from variable join station on station.station_code=variable.station_code;",
+            # "select distinct filename from variable;",
+            "select distinct filename from variable join station on station.station_code=variable.station_code and not exists (select * from characteristic where var_id=variable.var_id and ct_type='Fraction');",
+            # "select distinct filename from variable join station on station.station_code=variable.station_code;",
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply tests/io/test_ebas_file_index.py]: done

File: tests/io/test_ebas_file_index.py [hunk 16]
```diff
@@ -35,11 +35,15 @@ def test_EbasSQLRequest__var2sql_error():
     [
         (
             {},
-            "select distinct filename from variable join station on station.station_code=variable.station_code;",
+            # "select distinct filename from variable;",
```
[the remark this answers is not in the document]
Comment (jgriesfeller) [reply tests/io/test_ebas_file_index.py]: done

**Review by avaldebe — APPROVED** [review]
🚀 🎆 👏🏼

**Pull request merged**